File: kosql-execute-async-java/src/main/java/com/github/jacokoo/kosql/execute/async/java/Dao.java
package com.github.jacokoo.kosql.execute.async.java;

import com.github.jacokoo.kosql.compose.Column;
import com.github.jacokoo.kosql.compose.Entity;
import com.github.jacokoo.kosql.compose.Order;
import com.github.jacokoo.kosql.compose.Table;
import com.github.jacokoo.kosql.compose.expression.Count;
import com.github.jacokoo.kosql.compose.expression.Exp;
import com.github.jacokoo.kosql.compose.expression.Operator;
import com.github.jacokoo.kosql.compose.statement.*;
import com.github.jacokoo.kosql.compose.typesafe.Column1;
import io.vertx.core.Future;
import kotlin.Pair;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public interface Dao extends Executor, Operator {

    /**
     * Insert an entity to db
     * @param entity the entity to be inserted
     * @param <T> primary key type
     * @return null if insert failed, otherwise the generated primary key value
     */
    default <T> Future<T> save(Entity<T> entity) {
        var table = entity.innerTable();
        var key = table.primaryKey();
        var columns = table.getColumns()
            .stream().filter(it -> !it.getAutoIncrement() || !it.getType().getNullValue().equals(entity.get(it.getName())))
            .collect(Collectors.toList());

        var values = new Values(columns.stream().map(it -> entity.get(it.getName())).collect(Collectors.toList()));
        var insert = new InsertData<>(table, new Columns(columns), List.of(values), null, false);
        return execute(new InsertEnd<>(insert)).map(it -> {
            // consider throw an exception here
            if (it.getSecond() != 1) return null;
            if (key.getAutoIncrement()) entity.set(key.getName(), it.getFirst());
            return it.getFirst();
        });
    }

    /**
     * Retrieve data row by primary key
     * @param table the specified table
     * @param id primary key value
     * @return null if not found, otherwise the entity
     */
    default <T, E extends Entity<T>, R extends Table<T, E>> Future<E> byId(R table, T id) {
        var columns = new Columns(table.getInner().getColumns());
        var exp = EQ(table.primaryKey(), id);
        var data = new SelectData<>(columns, table, List.of(), exp, List.of(), null, List.of(), null, null);
        return query(new SelectEnd<>(data)).fetch(table).map(it -> it.isEmpty() ? null : it.get(0));
    }

    default <T, E extends Entity<T>, R extends Table<T, E>> Future<Integer> count(R table, Exp<?> exp) {
        var col = new Count<Object>("");
        var cs = new Column1<Integer>(col, List.of(col));
        var data = new SelectData<>(cs, table, List.of(), exp, List.of(), null, List.of(), null, null);
        return query1(new SelectEnd<>(data)).fetchValue();
    }

    default <T, E extends Entity<T>, R extends Table<T, E>> Future<List<E>> fetch(R table, Exp<?> exp, Pair<Column<?>, Order>...orders) {
        var cs = new Columns(table.getInner().getColumns());
        var data = new SelectData<>(cs, table, List.of(), exp, List.of(), null, Arrays.asList(orders), null, null);
        return query(new SelectEnd<>(data)).fetch(table);
    }

    default <T, E extends Entity<T>, R extends Table<T, E>> Future<Boolean> delete(R table, T key) {
        var data = new DeleteData(List.of(table), null, List.of(), EQ(table.primaryKey(), key));
        return execute(new DeleteEnd(data)).map(it -> it == 1);
    }

    default <T, E extends Entity<T>, R extends Table<T, E>> Future<Integer> delete(R table, Exp<?> exp) {
        var data = new DeleteData(List.of(table), null, List.of(), exp);
        return execute(new DeleteEnd(data));
    }
}

File: kosql-execute-async/src/main/kotlin/com/github/jacokoo/kosql/execute/async/dao.kt
package com.github.jacokoo.kosql.execute.async

import com.github.jacokoo.kosql.compose.Column
import com.github.jacokoo.kosql.compose.Entity
import com.github.jacokoo.kosql.compose.Order
import com.github.jacokoo.kosql.compose.Table
import com.github.jacokoo.kosql.compose.expression.Count
import com.github.jacokoo.kosql.compose.expression.Exp
import com.github.jacokoo.kosql.compose.expression.Operator
import com.github.jacokoo.kosql.compose.statement.*
import com.github.jacokoo.kosql.compose.typesafe.Column1

interface Dao: Executor, Operator {
    suspend fun <T> Entity<T>.save(): T {
        val table = innerTable()
        val key = table.primaryKey()
        val columns = table.columns.filter { !it.autoIncrement || this[it.name] != it.type.nullValue }

        val values = Values(columns.map { this[it.name] })
        val id = execute(InsertEnd(InsertData(table, Columns(columns), listOf(values))))

        if (key.autoIncrement) this[key.name] = id
        return id
    }

    suspend fun <T, E: Entity<T>> List<E>.save(useManualId: Boolean = false): List<E> {
        assert(isNotEmpty())

        val table = first().innerTable()
        var columns = table.columns
        val key = table.primaryKey()
        if (!useManualId && key.autoIncrement) columns = columns.filter { it != key }
        val values = map { e -> Values(columns.map { e[it.name] }) }
        val ids = execute(BatchInsertEnd(InsertData(table, Columns(columns), values, null, true)))

        assert(size == ids.size)
        if (!useManualId && key.autoIncrement) {
            forEachIndexed { i, it -> it[key.name] = ids[i] }
        }
        return this
    }

    suspend fun <T, E: Entity<T>, R: Table<T, E>> R.byId(t: T): E?
            = execute(SelectEnd(SelectData(Columns(inner.columns), this, expression = primaryKey() EQ t)), this).firstOrNull()

    suspend fun <T, R: Table<T, Entity<T>>> R.count(exp: Exp<*>? = null): Int
            = SelectEnd(SelectData(Column1(Count<Any>()), this, expression = exp)).fetchValue()

    suspend fun <T, E: Entity<T>, R: Table<T, E>> R.fetch(exp: Exp<*>? = null, vararg orders: Pair<Column<*>, Order>): List<E>
            = execute(SelectEnd(SelectData(Columns(inner.columns), this, expression = exp, orderBy = orders.toList())), this)

    suspend fun <T, R: Table<T, Entity<T>>> R.delete(t: T): Boolean =
        execute(DeleteEnd(DeleteData(deletes = listOf(this), expression = primaryKey() EQ t))) == 1

    suspend fun <T, R: Table<T, Entity<T>>> R.delete(exp: Exp<*>? = null): Int =
        execute(DeleteEnd(DeleteData(deletes = listOf(this), expression = exp)))

}
